lib/shrine/plugins/derivation_endpoint.rb

```diff
@@ -0,0 +1,1295 @@
+    # from derivation name, arguments and source file id. You can override that
+    # with the `:disposition` and `:filename` options:
+    #
+    #     plugin :derivation_endpoint,
+    #       disposition: -> { name == :thumbnail ? "inline" : "attachment" },
+    #       filename:    -> { [name, *args].join("-") }
+    #
+    # When the user opens the link in the browser, an `inline` disposition will
+    # tell the browser to render the file if possible, while `attachment`
+    # disposition will force download.
+    #
+    # The `:filename` and `:disposition` options can also be set per URL:
+    #
+    #     uploaded_file.derivation_url(:pdf, disposition: "attachment", filename: "custom-filename")
+    #     #=> "/thumbnail/eyJpZCI6ImZvbyIsInN?disposition=attachment&filename=custom-filename&signature=..."
+    #
+    # ## Uploading
+    #
+    # By default the derivation from a source file will be called each time
+    # it's requested. However, you can cache the derivatives to a storage by
+    # setting `:upload` to `true`. On first request the generated derivative
+    # will be uploaded to storage, and then on subsequent requests the already
+    # uploaded derivative will be served from the storage.
+    #
+    #     plugin :derivation_endpoint, upload: true
+    #
+    # The derivative will be uploaded to `<source id>/<name>-<args>` by default,
+    # and can be changed via `:upload_location`:
+    #
+    #     plugin :derivation_endpoint, upload: true, upload_location: -> {
+    #       # e.g. "derivatives/9a7d1bfdad24a76f9cfaff137fe1b5c7/thumbnail-1000-800"
+    #       ["derivatives", File.basename(source.id, ".*"), [name, *args].join("-")].join("/")
+    #     }
+    #
+    # Since the default upload location won't have any file extension, the
+    # derivation response won't know the appropriate `Content-Type` header
+    # value, so it will be set to `application/octet-stream`. It's recommended
+    # to use the `:type` option to set the appropriate `Content-Type` value.
+    #
+    # The target storage used is the same as for the source uploaded file, but
+    # it can be changed via `:upload_storage`:
+    #
+    #     plugin :derivation_endpoint, upload: true,
+    #                                  upload_storage: :thumbnail_storage
+    #
+    # Additional storage-specific upload options can be passed via
+    # `:upload_options`:
+    #
+    #     plugin :derivation_endpoint, upload: true,
+    #                                  upload_options: { acl: "public-read" }
+    #
+    # ### Redirecting
+    #
+    # The derivative content will be served through the endpoint by default.
+    # However, you can configure the endpoint to redirect to the uploaded
+    # derivative on the storage:
+    #
+    #     plugin :derivation_endpoint, upload: true,
+    #                                  upload_redirect: true
+    #
+    # In that case additional storage-specific URL options can also be passed in:
```

Comment:
Since we are redirecting the URL to the storage, additional storage-specific ....